hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+  std::vector<uint8_t> out(in.size(), 0);
+
+  // Iterate through each 4 bit chunk of the data and apply the appropriate SBOX
+  for (int i = 0; i < bit_width / 4; ++i) {
+    uint8_t sbox_in, sbox_out;
+
+    sbox_in = in[i / 2];
+
+    if (i % 2) {
+      sbox_in >>= 4;
+    } else {
+      sbox_in &= 0xf;
+    }
+
+    if (invert) {
+      sbox_out = PRESENT_SBOX4_INV[sbox_in];
+    } else {
+      sbox_out = PRESENT_SBOX4[sbox_in];
+    }
+
+    if (i % 2) {
+      out[i / 2] |= sbox_out << 4;
+    } else {
+      out[i / 2] |= sbox_out;
+    }
+  }
+
+  // Where bit_width is not a multiple of 4 copy over the remaining bits
+  if (bit_width % 4) {
+    if (bit_width % 8 >= 4) {
+      // Final bits are in top half of final byte
+      out[bit_width / 8] =
+          (out[bit_width / 8] & 0xf) | (in[bit_width / 8] & 0xf0);
+    } else {
+      // Final bits are in bottom half of final byte
+      out[bit_width / 8] = in[bit_width / 8] & 0xf;
+    }
+  }
+
+  return out;
+}
+
+// Mirror bits from incoming byte vector
+static std::vector<uint8_t> scramble_flip_layer(const std::vector<uint8_t> &in,
+                                                uint32_t bit_width) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  std::vector<uint8_t> out(in.size(), 0);
+
+  for (int i = 0; i < bit_width; ++i) {
+    or_vector_bit(out, bit_width - i - 1, read_vector_bit(in, i));
+  }
+
+  return out;
+}
+
+// Apply butterfly to incoming byte vector. Even bits are placed in the lower
+// half of the output, odd bits are placed in the upper half of the output.
+static std::vector<uint8_t> scramble_perm_layer(const std::vector<uint8_t> &in,
+                                                uint32_t bit_width,
+                                                bool invert) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  std::vector<uint8_t> out(in.size(), 0);
+
+  for (int i = 0; i < bit_width / 2; ++i) {
+    if (invert) {
+      or_vector_bit(out, i * 2, read_vector_bit(in, i));
+      or_vector_bit(out, i * 2 + 1, read_vector_bit(in, i + (bit_width / 2)));
+    } else {
+      or_vector_bit(out, i, read_vector_bit(in, i * 2));
+      or_vector_bit(out, i + (bit_width / 2), read_vector_bit(in, i * 2 + 1));
+    }
+  }
+
+  if (bit_width % 2) {
+    // Where bit_width isn't even final bit is copied across to the same
+    // position
+    or_vector_bit(out, bit_width - 1, read_vector_bit(in, bit_width - 1));
+  }
+
+  return out;
+}
+
+// Apply a full set of subsitution/permutation rounds for encrypt to the
+// incoming byte vector
+static std::vector<uint8_t> scramble_subst_perm_enc(
+    const std::vector<uint8_t> &in, const std::vector<uint8_t> &key,
+    uint32_t bit_width, uint32_t num_rounds) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  assert(key.size() >= ((bit_width + 7) / 8));
+  std::vector<uint8_t> state(in);
+
+  for (int i = 0; i < num_rounds; ++i) {
+    for (int j = 0; j < in.size(); ++j) {
+      state[j] = state[j] ^ key[j];
+    }
```

Comment:
Easier: Initialise `state` as a copy of `key` and then move this `for` loop after the call to `scramble_perm_layer`. Then you can get rid of the final state update below.

Similarly for the decode function.

Reply:
Will that work? `state` is initialised with `in` so this will do an initial XORing of `in` with `key` before any of scramble_X_layer calls below. If we initialise with `key` we loose that initial XOR with `in`. Indeed `in` wouldn't factor into the function at all, which can't be correct!

I could initialise `state` as `key` XOR `in` but then I'll just have an extra loop at the beginning of the function rather than the end.

Reply:
Oh, you're completely right, sorry! For some reason, I thought that we were initialising `state` with zeros.

I wonder whether it would make sense to define a wrapper for the XOR update?
```c++
static void xor_update(std::vector<uint8_t> &acc, const std::vector<uint8_t> &upd) {
  assert(acc.size() == upd.size());
  for (size_t i = 0; i < acc.size(); ++i) {
    acc[i] ^= upd[i];
  }
}
```
then you could use it here as:
```c++
  xor_update(state, key);
```
That way we can at least get rid of the repeated loop in the code.

Reply:
Done, though a slightly different function signature (takes two vectors in, XORs them, returns new vector) as there was some XORing happening in `scramble_decypt_data` and `scramble_encrypt_data` where the destination was different from the source. I could have modified them so they worked with the proposed `xor_update` but I felt `dst = xor_vectors(src1, src2)` was an easier understand line of code.